### PR TITLE
feat: add support for algebraic effects in `upcast`

### LINKED ISSUE #4282: Better upcast for set effects
Currently `upcast` requires a pure function.

This is fine for the binary effects, but prevents the set effects from upcasting from `{Print}` to `{Print, Throw}`.

I think we could implement a safe set upcast with something like `upcast(a -> b \ e): (a -> b \ {e, f})`.

### PR DESCRIPTION
**Pull request by jaschdoc**
Related to https://github.com/flix/flix/issues/4264 https://github.com/flix/flix/issues/4486
Closes #4282

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/phase/Safety.scala [hunk 1]
```diff
@@ -44,255 +45,267 @@ object Safety {
   /**
     * Performs safety and well-formedness checks on the given definition `def0`.
     */
-  private def visitDef(def0: Def): List[CompilationMessage] = visitExp(def0.impl.exp)
+  private def visitDef(def0: Def)(implicit flix: Flix): List[CompilationMessage] = {
+    val rigidityEnv = def0.spec.tparams.map(_.sym).foldLeft(RigidityEnv.empty) {
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/Safety.scala:49]: `renv` is the standard variable name for this.

File: main/src/ca/uwaterloo/flix/language/phase/Safety.scala [hunk 2]
```diff
@@ -44,255 +45,267 @@ object Safety {
...
 
   /**
     * Performs safety and well-formedness checks on the given expression `exp0`.
     */
-  private def visitExp(exp0: Expression): List[CompilationMessage] = exp0 match {
-    case Expression.Unit(_) => Nil
+  private def visitExp(exp0: Expression, rigidityEnv: RigidityEnv)(implicit flix: Flix): List[CompilationMessage] = {
 
-    case Expression.Null(_, _) => Nil
+    def vstExp(exp1: Expression): List[CompilationMessage] = exp1 match {
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/Safety.scala:61]: Inner visitors are often called `visit`.

File: main/src/ca/uwaterloo/flix/language/phase/Safety.scala [hunk 3]
```diff
@@ -44,255 +45,267 @@ object Safety {
...
+    visitExp(def0.impl.exp, rigidityEnv)
+  }
+
 
   /**
     * Performs safety and well-formedness checks on the given expression `exp0`.
     */
-  private def visitExp(exp0: Expression): List[CompilationMessage] = exp0 match {
-    case Expression.Unit(_) => Nil
+  private def visitExp(exp0: Expression, rigidityEnv: RigidityEnv)(implicit flix: Flix): List[CompilationMessage] = {
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/Safety.scala:59]: `renv`

File: main/src/ca/uwaterloo/flix/language/phase/Safety.scala (hunk 2 above)
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/Safety.scala:61]: `exp1` has shadows below. Can we do `e0` for the top expression, then `exp0` for the inner one?

File: main/src/ca/uwaterloo/flix/language/phase/Safety.scala [hunk 4]
```diff
@@ -309,10 +322,14 @@ object Safety {
...
-    * the purity of a function is being cast from `pure` -> `ef` -> `impure`.
+    * the purity of the expression is being cast from `pure` -> `ef` -> `impure`.
+    *
+    * OR
+    *
+    * the effect set of the expression is a subset of the effect set being cast to.
     *
     */
-  private def isSubTypeOf(tpe1: Type, tpe2: Type): Boolean = (tpe1.baseType, tpe2.baseType) match {
+  private def isSubTypeOf(tpe1: Type, tpe2: Type, rigidityEnv: RigidityEnv)(implicit flix: Flix): Boolean = (tpe1.baseType, tpe2.baseType) match {
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/Safety.scala:332]: renv

File: main/test/flix/Test.Exp.Upcast.flix [hunk 5]
```diff
@@ -128,4 +128,30 @@ namespace Test/Exp/Upcast {
...
+        let f = (_: ##java.lang.Object) -> newStringBuilder();
+        let g = (_: ##java.lang.StringBuilder) -> newObject();
+        let _ =
+            if (true)
+                upcast f
+            else
+                g;
+        ()
+
 }
```
Comment (mlutze) [inline main/test/flix/Test.Exp.Upcast.flix:157]: Can we add some more complex set formulas?

By the way, I think an easy test structure would be something like this:

```
(upcast (???: SomeType)): TargetType
```

**Review by mlutze — CHANGES_REQUESTED** [review]

File: main/test/flix/Test.Exp.Upcast.flix (hunk 5 above)
Comment (jaschdoc) [reply main/test/flix/Test.Exp.Upcast.flix:157]: That doesn't work at the moment as we can only upcast effecst on functions. I'll get around later to upcasting expressions (in a new pr) :)

**Review by mlutze — APPROVED** [review]

**Pull request merged**